Assets/MixedRealityToolkit.Services/InputAnimation/InputAnimation.cs

```diff
@@ -0,0 +1,710 @@
+
+        /// <summary>
+        /// Maximum duration of all animations curves.
+        /// </summary>
+        [SerializeField]
+        private float duration = 0.0f;
+        public float Duration => duration;
+
+        public class PoseCurves
+        {
+            readonly public AnimationCurve PositionX = new AnimationCurve();
+            readonly public AnimationCurve PositionY = new AnimationCurve();
+            readonly public AnimationCurve PositionZ = new AnimationCurve();
+            readonly public AnimationCurve RotationX = new AnimationCurve();
+            readonly public AnimationCurve RotationY = new AnimationCurve();
+            readonly public AnimationCurve RotationZ = new AnimationCurve();
+            readonly public AnimationCurve RotationW = new AnimationCurve();
+        }
+
+        [SerializeField]
+        private AnimationCurve handTrackedCurveLeft;
+        [SerializeField]
+        private AnimationCurve handTrackedCurveRight;
+        [SerializeField]
+        private AnimationCurve handPinchCurveLeft;
+        [SerializeField]
+        private AnimationCurve handPinchCurveRight;
+        [SerializeField]
+        private Dictionary<TrackedHandJoint, PoseCurves> handJointCurvesLeft;
+        [SerializeField]
+        private Dictionary<TrackedHandJoint, PoseCurves> handJointCurvesRight;
+        [SerializeField]
+        private PoseCurves cameraCurves;
+        public PoseCurves CameraCurves => cameraCurves;
+
+        /// <summary>
+        /// Number of markers in the animation.
+        /// </summary>
+        [SerializeField]
+        private List<InputAnimationMarker> markers;
+        public int markerCount => markers.Count;
+
+        internal class CompareMarkers : IComparer<InputAnimationMarker>
+        {
+            public int Compare(InputAnimationMarker a, InputAnimationMarker b)
+            {
+                return a.time.CompareTo(b.time);
+            }
+        }
+
+        public InputAnimation()
+        {
+            handTrackedCurveLeft = new AnimationCurve();
+            handTrackedCurveRight = new AnimationCurve();
+            handPinchCurveLeft = new AnimationCurve();
+            handPinchCurveRight = new AnimationCurve();
+            handJointCurvesLeft = new Dictionary<TrackedHandJoint, PoseCurves>();
+            handJointCurvesRight = new Dictionary<TrackedHandJoint, PoseCurves>();
+            cameraCurves = new PoseCurves();
+
+            markers = new List<InputAnimationMarker>();
+        }
+
+        public bool TryGetHandJointCurves(Handedness handedness, TrackedHandJoint joint, out PoseCurves curves)
+        {
+            if (handedness == Handedness.Left)
+            {
+                return handJointCurvesLeft.TryGetValue(joint, out curves);
+            }
+            else if (handedness == Handedness.Right)
+            {
+                return handJointCurvesRight.TryGetValue(joint, out curves);
+            }
+            curves = new PoseCurves();
```

Comment:
Nit: What happens here if we just null out curves (i.e. no need to allocate for what is an error case right?)

Reply:
Yes, returning null now. I was using a struct for PoseCurves before which means null wouldn't work (and iirc would be stack-allocated), but for various reasons it's easier to use a class for the PoseCurves collection.